borg/repository.py

```diff
@@ -188,10 +191,28 @@ def commit(self, save_space=False):
         self.write_index()
         self.rollback()
 
-    def open_index(self, transaction_id):
+    def open_index(self, transaction_id, auto_recover=True):
         if transaction_id is None:
             return NSIndex()
-        return NSIndex.read((os.path.join(self.path, 'index.%d') % transaction_id).encode('utf-8'))
+        index_path = os.path.join(self.path, 'index.%d' % transaction_id).encode('utf-8')
+        try:
+            return NSIndex.read(index_path)
+        except RuntimeError as re:
+            assert str(re) == 'hashindex_read failed'  # everything else means we're in *deep* trouble
+            # corrupted index file, need to replay segments
+            try:
+                os.unlink(os.path.join(self.path, 'hints.%d' % transaction_id))
+                os.unlink(os.path.join(self.path, 'index.%d' % transaction_id))
+            except OSError as e:
+                raise InternalOSError from e
+            if not auto_recover:
+                raise
```

Comment:
aren't the unlink commands above already part of the auto recovery? if so, move these 2 lines upwards to above the unlinks?